cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/GoalOptimizer.java

```diff
@@ -117,6 +117,52 @@ public GoalOptimizer(KafkaCruiseControlConfig config,
     _proposalComputationTimer = dropwizardMetricRegistry.timer(MetricRegistry.name("GoalOptimizer", "proposal-computation-timer"));
   }
 
+  /**
+   * Generate unique goal priority for each of proposal precomputing thread.
+   */
+  private void populateGoalByPriorityForPrecomputing() {
+    List<List<Goal>> shuffledGoalList = new ArrayList<>();
+    // Guarantee one thread is working on the original goal priority.
+    shuffledGoalList.add(new ArrayList<>(_goalsByPriority.values()));
+
+    Integer goalCount = _goalsByPriority.values().size();
+    Integer goalIndex = goalCount - 2;
+    Integer numShuffledGoalToGenerate = numProposalComputingThreads() - 1;
+    while (numShuffledGoalToGenerate > 0) {
```

Comment:
What happens when `numShuffledGoalToGenerate` = `3` and `goalCount` = `2`?

Reply:
in this case, _numPrecomputingThreads will be min(3, 2!)=2, numProposalComputingThreads() will be 2, numShuffledGoalsToGenerate will be 1. and while loop will execute once, the result will be the two possible combination of goal priority.